engine/transaction/draft/outputs/paymail.go

```diff
@@ -59,7 +66,49 @@ func (p *Paymail) createBsvPaymailOutput(output *paymail.PaymentOutput, referenc
 			Paymail: &transaction.PaymailAnnotation{
 				Receiver:  p.To,
 				Reference: reference,
+				Sender:    from,
 			},
 		},
 	}, nil
 }
+
+func (p *Paymail) sender(ctx evaluation.Context) (string, error) {
+	if p.From != nil {
+		err := p.validateProvidedSenderPaymail(ctx)
+		if err != nil {
+			return "", err
+		}
+
+		return *p.From, nil
+	}
+	return p.defaultSenderAddress(ctx)
+}
+
+func (p *Paymail) validateProvidedSenderPaymail(ctx evaluation.Context) error {
+	var sender = *p.From
+	_, err := ctx.Paymail().GetSanitizedPaymail(sender)
+	if err != nil {
+		return txerrors.ErrSenderPaymailAddressIsInvalid.Wrap(err)
+	}
+	ownsPaymail, err := ctx.PaymailAddressService().HasPaymailAddress(ctx, ctx.XPubID(), sender)
+	if errors.Is(err, spverrors.ErrCouldNotFindPaymail) {
+		return txerrors.ErrSenderPaymailAddressIsInvalid.Wrap(err)
+	}
+	if err != nil {
+		return spverrors.Wrapf(err, "failed to check if paymail %s belongs to xpub %s", sender, ctx.XPubID())
+	}
+
+	if !ownsPaymail {
+		return txerrors.ErrSenderPaymailAddressIsInvalid
```

Comment:
I don't know if it's necessary, but, maybe, we could distinguish the situation when `ErrCouldNotFindPaymail` and `!ownsPaymail`.

But, perhaps it's ok because of the wrapping, so for the not-find situation we can always do `errors.Is(err, ErrCouldNotFindPaymail)`

Reply:
From the client point of view, I'm not sure if this is really helpful information that he is trying to use someone else paymail.

Reply:
Ok. Fine for me 👍